From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Narayan Reddy <narayanr@nvidia.com>
Date: Sun, 24 Oct 2021 02:28:06 +0530
Subject: [PATCH 030/245] NVIDIA: SAUCE: net: phy: aqr113c: Enable Wake on boot
 configuration.

Enable wake on boot configuration. This feature
configures the Wake on LAN settings in PHY where
PHY detects the magic packet and generates an
interrupt to boot the device.

Bug 3270211

Change-Id: I4ee6d0ec7f9a1786337eefa8ce75b8f8a896a093
Signed-off-by: Narayan Reddy <narayanr@nvidia.com>
Signed-off-by: Revanth Kumar Uppala <ruppala@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2909744
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
Tested-by: Abhilash G <abhilashg@nvidia.com>

diff --git a/drivers/net/phy/aquantia/aquantia_main.c b/drivers/net/phy/aquantia/aquantia_main.c
index a69bb0a16638e..3f8d2a614690c 100644
--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -5,6 +5,7 @@
  * Author: Shaohui Xie <Shaohui.Xie@freescale.com>
  *
  * Copyright 2015 Freescale Semiconductor, Inc.
+ * Copyright (c) 2021-2023, NVIDIA CORPORATION.  All rights reserved.
  */
 
 #include <linux/kernel.h>
@@ -13,6 +14,7 @@
 #include <linux/bitfield.h>
 #include <linux/of.h>
 #include <linux/phy.h>
+#include <linux/netdevice.h>
 
 #include "aquantia.h"
 
@@ -56,6 +58,7 @@
 #define MDIO_AN_VEND_PROV			0xc400
 #define MDIO_AN_VEND_PROV_1000BASET_FULL	BIT(15)
 #define MDIO_AN_VEND_PROV_1000BASET_HALF	BIT(14)
+#define MDIO_AN_VEND_PROV_AQRATE_DWN_SHFT_CAP	BIT(12)
 #define MDIO_AN_VEND_PROV_5000BASET_FULL	BIT(11)
 #define MDIO_AN_VEND_PROV_2500BASET_FULL	BIT(10)
 #define MDIO_AN_VEND_PROV_EXC_PHYID_INFO	BIT(6)
@@ -113,8 +116,26 @@
 #define MDIO_AN_RX_VEND_STAT3			0xe832
 #define MDIO_AN_RX_VEND_STAT3_AFR		BIT(0)
 
+#define MDIO_AN_LD_LOOP_TIMING_ABILITY	BIT(0)
+#define MDIO_MMD_AN_WOL_ENABLE		BIT(6)
+#define MDIO_AN_VEND_MASK		0xF0FF
+
+/* MDIO_MMD_C22EXT */
+#define MDIO_C22EXT_MAGIC_PKT_PATTREN_0_2_15		0xc339
+#define MDIO_C22EXT_MAGIC_PKT_PATTREN_16_2_31		0xc33a
+#define MDIO_C22EXT_MAGIC_PKT_PATTREN_32_2_47		0xc33b
+#define MDIO_C22EXT_GBE_PHY_RSI1_CTRL6			0xc355
+#define MDIO_C22EXT_GBE_PHY_RSI1_CTRL7			0xc356
+#define MDIO_C22EXT_GBE_PHY_RSI1_CTRL8			0xc357
+#define MDIO_C22EXT_GBE_PHY_SGMII_TX_INT_MASK1		0xf420
 #define MDIO_C22EXT_GBE_PHY_SGMII_TX_ALARM1		0xec20
 
+#define MDIO_C22EXT_RSI_WAKE_UP_FRAME_DETECTION		BIT(0)
+#define MDIO_C22EXT_RSI_MAGIC_PKT_FRAME_DETECTION	BIT(0)
+#define MDIO_C22EXT_RSI_WOL_FCS_MONITOR_MODE		BIT(15)
+#define MDIO_C22EXT_SGMII0_WAKE_UP_FRAME_MASK		BIT(4)
+#define MDIO_C22EXT_SGMII0_MAGIC_PKT_FRAME_MASK		BIT(5)
+
 /* Sleep and timeout for checking if the Processor-Intensive
  * MDIO operation is finished
  */
@@ -123,6 +144,17 @@
 
 #define VEND1_SEC_INGRESS_CNTRL_REG1		0x7001
 
+#define VEND1_GLOBAL_SYS_CONFIG_SGMII		(BIT(0) | BIT(1) | BIT(3))
+#define VEND1_GLOBAL_SYS_CONFIG_XFI		BIT(8)
+
+#define BIT_SHIFT_8 8
+#define MAC_ADDRESS_BYTE_0 0
+#define MAC_ADDRESS_BYTE_1 1
+#define MAC_ADDRESS_BYTE_2 2
+#define MAC_ADDRESS_BYTE_3 3
+#define MAC_ADDRESS_BYTE_4 4
+#define MAC_ADDRESS_BYTE_5 5
+
 static int aqr107_get_sset_count(struct phy_device *phydev)
 {
 	return AQR107_SGMII_STAT_SZ;
@@ -312,11 +344,11 @@ static irqreturn_t aqr_handle_interrupt(struct phy_device *phydev)
 
 		/* Restore the SERDES/System Interface back to the XFI mode */
 		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
-				    VEND1_GLOBAL_SYS_CONFIG_100M, VEND1_GLOBAL_SYS_CONFIG_XFI);
+				    VEND1_GLOBAL_CFG_100M, VEND1_GLOBAL_SYS_CONFIG_XFI);
 		if (ret < 0)
 			return IRQ_NONE;
 		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
-				    VEND1_GLOBAL_SYS_CONFIG_1G, VEND1_GLOBAL_SYS_CONFIG_XFI);
+				    VEND1_GLOBAL_CFG_1G, VEND1_GLOBAL_SYS_CONFIG_XFI);
 		if (ret < 0)
 			return IRQ_NONE;
 		/* restart auto-negotiation */
@@ -805,7 +837,29 @@ static int aqr_gen1_config_init(struct phy_device *phydev)
 {
 	struct aqr107_priv *priv = phydev->priv;
 	u32 led_idx;
-	int ret;
+	int ret, err;
+
+	/* Check that the PHY interface type is compatible */
+	if (phydev->interface != PHY_INTERFACE_MODE_SGMII &&
+	    phydev->interface != PHY_INTERFACE_MODE_2500BASEX &&
+	    phydev->interface != PHY_INTERFACE_MODE_XGMII &&
+	    phydev->interface != PHY_INTERFACE_MODE_USXGMII &&
+	    phydev->interface != PHY_INTERFACE_MODE_10GKR &&
+	    phydev->interface != PHY_INTERFACE_MODE_10GBASER)
+		return -ENODEV;
+
+	WARN(phydev->interface == PHY_INTERFACE_MODE_XGMII,
+	     "Your devicetree is out of date, please update it. The AQR107 family doesn't support XGMII, maybe you mean USXGMII.\n");
+
+	ret = aqr_wait_reset_complete(phydev);
+	if (!ret) {
+		/* The PHY might work without a firmware image, so only build a
+		 * fingerprint if the firmware was initialized.
+		 */
+		ret = aqr_build_fingerprint(phydev);
+		if (ret)
+			return ret;
+	}
 
 	/* Advertize flow control */
 	linkmode_set_bit(SUPPORTED_Pause, phydev->supported);
@@ -827,30 +881,29 @@ static int aqr_gen1_config_init(struct phy_device *phydev)
 	if (err < 0)
 		return err;
 
-	/* Check that the PHY interface type is compatible */
-	if (phydev->interface != PHY_INTERFACE_MODE_SGMII &&
-	    phydev->interface != PHY_INTERFACE_MODE_1000BASEKX &&
-	    phydev->interface != PHY_INTERFACE_MODE_2500BASEX &&
-	    phydev->interface != PHY_INTERFACE_MODE_XGMII &&
-	    phydev->interface != PHY_INTERFACE_MODE_USXGMII &&
-	    phydev->interface != PHY_INTERFACE_MODE_10G_QXGMII &&
-	    phydev->interface != PHY_INTERFACE_MODE_10GKR &&
-	    phydev->interface != PHY_INTERFACE_MODE_10GBASER &&
-	    phydev->interface != PHY_INTERFACE_MODE_XAUI &&
-	    phydev->interface != PHY_INTERFACE_MODE_RXAUI)
-		return -ENODEV;
+	/* Configure Magic packet frame pattern (MAC address) */
+	err = phy_write_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_MAGIC_PKT_PATTREN_0_2_15,
+			    phydev->attached_dev->dev_addr[MAC_ADDRESS_BYTE_0] |
+			    (phydev->attached_dev->dev_addr[MAC_ADDRESS_BYTE_1] << BIT_SHIFT_8));
+	if (err < 0) {
+		phydev_err(phydev, "Error setting magic packet frame of 0/1st byte\n");
+		return err;
+	}
 
-	WARN(phydev->interface == PHY_INTERFACE_MODE_XGMII,
-	     "Your devicetree is out of date, please update it. The AQR107 family doesn't support XGMII, maybe you mean USXGMII.\n");
+	err = phy_write_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_MAGIC_PKT_PATTREN_16_2_31,
+			    phydev->attached_dev->dev_addr[MAC_ADDRESS_BYTE_2] |
+			    (phydev->attached_dev->dev_addr[MAC_ADDRESS_BYTE_3] << BIT_SHIFT_8));
+	if (err < 0) {
+		phydev_err(phydev, "Error setting magic packet frame of 2/3rd byte\n");
+		return err;
+	}
 
-	ret = aqr_wait_reset_complete(phydev);
-	if (!ret) {
-		/* The PHY might work without a firmware image, so only build a
-		 * fingerprint if the firmware was initialized.
-		 */
-		ret = aqr_build_fingerprint(phydev);
-		if (ret)
-			return ret;
+	err = phy_write_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_MAGIC_PKT_PATTREN_32_2_47,
+			    phydev->attached_dev->dev_addr[MAC_ADDRESS_BYTE_4] |
+			    (phydev->attached_dev->dev_addr[MAC_ADDRESS_BYTE_5] << BIT_SHIFT_8));
+	if (err < 0) {
+		phydev_err(phydev, "Error setting magic packet frame of 4/5th byte\n");
+		return err;
 	}
 
 	ret = aqr107_set_downshift(phydev, MDIO_AN_VEND_PROV_DOWNSHIFT_DFLT);
@@ -1228,6 +1281,131 @@ static int aqr107_probe(struct phy_device *phydev)
 	return aqr_hwmon_probe(phydev);
 }
 
+static int aqr113c_wol_settings(struct phy_device *phydev, bool enable)
+{
+	u16 val;
+	int ret = 0;
+
+	if (enable) {
+		/* Disables all advertised speeds except for the WoL
+		 * speed (100BASE-TX FD or 1000BASE-T)
+		 * This is set as per the APP note from Marvel
+		 */
+		val = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_10GBT_CTRL);
+		val |= MDIO_AN_LD_LOOP_TIMING_ABILITY;
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_10GBT_CTRL, val);
+		if (ret < 0)
+			return ret;
+		val = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_VEND_PROV);
+		val = (val & MDIO_AN_VEND_MASK) |
+		      (MDIO_AN_VEND_PROV_AQRATE_DWN_SHFT_CAP | MDIO_AN_VEND_PROV_1000BASET_FULL);
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_VEND_PROV, val);
+		if (ret < 0)
+			return ret;
+
+		/* Enable the magic frame and wake up frame detection for the PHY */
+		val = phy_read_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL6);
+		val |= MDIO_C22EXT_RSI_WAKE_UP_FRAME_DETECTION;
+		ret = phy_write_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL6, val);
+		if (ret < 0)
+			return ret;
+		val = phy_read_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL7);
+		val |= MDIO_C22EXT_RSI_MAGIC_PKT_FRAME_DETECTION;
+		ret = phy_write_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL7, val);
+		if (ret < 0)
+			return ret;
+
+		/* Set the WoL enable bit */
+		val = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_RESVD_VEND_PROV);
+		val |= MDIO_MMD_AN_WOL_ENABLE;
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_RESVD_VEND_PROV, val);
+		if (ret < 0)
+			return ret;
+
+		/* Set the WoL INT_N trigger bit */
+		val = phy_read_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL8);
+		val |= MDIO_C22EXT_RSI_WOL_FCS_MONITOR_MODE;
+		ret = phy_write_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL8, val);
+		if (ret < 0)
+			return ret;
+
+		/* Enable Interrupt INT_N Generation at pin level */
+		val = phy_read_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_SGMII_TX_INT_MASK1);
+		val |= MDIO_C22EXT_SGMII0_WAKE_UP_FRAME_MASK |
+		       MDIO_C22EXT_SGMII0_MAGIC_PKT_FRAME_MASK;
+		ret = phy_write_mmd(phydev, MDIO_MMD_C22EXT,
+				    MDIO_C22EXT_GBE_PHY_SGMII_TX_INT_MASK1, val);
+		if (ret < 0)
+			return ret;
+		val = phy_read_mmd(phydev, MDIO_MMD_VEND1, VEND1_GLOBAL_INT_STD_MASK);
+		val |= VEND1_GLOBAL_INT_STD_MASK_ALL;
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1, VEND1_GLOBAL_INT_STD_MASK, val);
+		if (ret < 0)
+			return ret;
+		val = phy_read_mmd(phydev, MDIO_MMD_VEND1, VEND1_GLOBAL_INT_VEND_MASK);
+		val |= VEND1_GLOBAL_INT_VEND_MASK_GBE;
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1, VEND1_GLOBAL_INT_VEND_MASK, val);
+		if (ret < 0)
+			return ret;
+
+		/* Set the system interface to SGMII */
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
+				    VEND1_GLOBAL_CFG_100M, VEND1_GLOBAL_SYS_CONFIG_SGMII);
+		if (ret < 0)
+			return ret;
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
+				    VEND1_GLOBAL_CFG_1G, VEND1_GLOBAL_SYS_CONFIG_SGMII);
+		if (ret < 0)
+			return ret;
+
+		/* restart auto-negotiation */
+		genphy_c45_restart_aneg(phydev);
+	} else {
+		/* Disable the WoL enable bit */
+		val = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_RESVD_VEND_PROV);
+		val &= ~MDIO_MMD_AN_WOL_ENABLE;
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_RESVD_VEND_PROV, val);
+		if (ret < 0)
+			return ret;
+
+		/* Restore the SERDES/System Interface back to the XFI mode */
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
+				    VEND1_GLOBAL_CFG_100M, VEND1_GLOBAL_SYS_CONFIG_XFI);
+		if (ret < 0)
+			return ret;
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
+				    VEND1_GLOBAL_CFG_1G, VEND1_GLOBAL_SYS_CONFIG_XFI);
+		if (ret < 0)
+			return ret;
+
+		/* restart auto-negotiation */
+		genphy_c45_restart_aneg(phydev);
+	}
+	return ret;
+}
+
+static void aqr113c_get_wol(struct phy_device *phydev, struct ethtool_wolinfo *wol)
+{
+	u16 val;
+
+	val = phy_read_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_RSI1_CTRL7);
+	if (val < 0)
+		return;
+
+	if (val & 0x1) {
+		wol->wolopts = WAKE_MAGIC;
+		wol->supported = WAKE_MAGIC;
+	}
+}
+
+static int aqr113c_set_wol(struct phy_device *phydev, struct ethtool_wolinfo *wol)
+{
+	if (wol->wolopts & WAKE_MAGIC)
+		return aqr113c_wol_settings(phydev, true);
+	else
+		return aqr113c_wol_settings(phydev, false);
+	return 0;
+}
 
 static struct phy_driver aqr_driver[] = {
 {
@@ -1373,6 +1551,8 @@ static struct phy_driver aqr_driver[] = {
 	.led_polarity_set = aqr_phy_led_polarity_set,
 	.inband_caps	= aqr_gen2_inband_caps,
 	.config_inband	= aqr_gen2_config_inband,
+	.get_wol	= &aqr113c_get_wol,
+	.set_wol	= &aqr113c_set_wol,
 },
 {
 	PHY_ID_MATCH_MODEL(PHY_ID_AQR405),
-- 
2.43.0

